File: DiKErnel/Src/DiKErnel.KernelWrapper.Json.Input/Public/JsonInputComposer.cpp
// Copyright (C) Stichting Deltares 2022. All rights reserved.
//
// This file is part of DiKErnel.
//
// DiKErnel is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, version 3.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
//
// All names, logos, and references to "Deltares" are registered trademarks of
// Stichting Deltares and remain full property of Stichting Deltares at all times.
// All rights reserved.

#include "JsonInputComposer.h"

#include <filesystem>
#include <fstream>

#include "EventRegistry.h"
#include "JsonInputAdapter.h"
#include "JsonInputParser.h"
#include "JsonSchemaDefinition.h"
#include "ValidationHelper.h"
#include "ValidationIssue.h"

namespace DiKErnel::KernelWrapper::Json::Input
{
    using namespace Core;
    using namespace Integration;
    using namespace nlohmann;
    using namespace json_schema;
    using namespace std;
    using namespace std::filesystem;
    using namespace Util;

    json_validator JsonInputComposer::_validator
    {
        json::parse(JSON_SCHEMA_DEFINITION)
    };

    bool JsonInputComposer::ValidateJson(
        const string& filePath)
    {
        vector<unique_ptr<ValidationIssue>> validationIssues;

        if (!exists(filePath))
        {
            validationIssues.emplace_back(make_unique<ValidationIssue>(ValidationIssueType::Error, "The provided input file does not exist"));
        }
        else
        {
            try
            {
                _validator.validate(json::parse(ifstream(filePath)), *_customErrorHandler);
            }
            catch (const json::parse_error& e)
            {
                validationIssues.emplace_back(make_unique<ValidationIssue>(ValidationIssueType::Error, e.what()));
            }
            catch (const invalid_argument& e)
            {
                validationIssues.emplace_back(make_unique<ValidationIssue>(ValidationIssueType::Error, e.what()));
            }
        }

        return ValidationHelper::RegisterValidationIssues(validationIssues);
    }

    unique_ptr<DataResult<ICalculationInput>> JsonInputComposer::GetInputDataFromJson(
        const string& filePath)
    {
        if (!exists(filePath))
        {
            EventRegistry::Register(make_unique<Event>("The provided input file does not exist", EventType::Error));
        }
        else
        {
            try
            {
                const auto jsonInputData = JsonInputParser::GetJsonInputData(filePath);
                return JsonInputAdapter::AdaptJsonInputData(*jsonInputData);
            }
            catch (const exception& e)
            {
                EventRegistry::Register(make_unique<Event>("An unhandled error occurred while composing calculation data from the Json input. See "
                                                           "stack trace for more information:\n" + static_cast<string>(e.what()),
                                                           EventType::Error));
            }
        }

        return make_unique<DataResult<ICalculationInput>>(EventRegistry::Flush());
    }
}

File: Src/DiKErnel.KernelWrapper.Json.Input/JsonInputComposer.cs
// Copyright (C) Stichting Deltares and State of the Netherlands 2023. All rights reserved.
//
// This file is part of DiKErnel.
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 3 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program; if not, see <https://www.gnu.org/licenses/>.
//
// All names, logos, and references to "Deltares" are registered trademarks of
// Stichting Deltares and remain full property of Stichting Deltares at all times.
// All rights reserved.

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using DiKErnel.Core.Data;
using DiKErnel.Integration.Helpers;
using DiKErnel.KernelWrapper.Json.Input.Data.Generic;
using DiKErnel.KernelWrapper.Json.Input.Data.Generic.Converters;
using DiKErnel.KernelWrapper.Json.Input.Properties;
using DiKErnel.Util;
using DiKErnel.Util.Validation;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Newtonsoft.Json.Schema;

namespace DiKErnel.KernelWrapper.Json.Input
{
    /// <summary>
    /// Class responsible for validating and parsing Json input.
    /// </summary>
    public static class JsonInputComposer
    {
        private static readonly JSchema schema;

        static JsonInputComposer()
        {
            using Stream validatorSchemaStream = new MemoryStream(Resources.SchemaDefinition);
            using var validatorSchemaReader = new StreamReader(validatorSchemaStream);
            schema = JSchema.Parse(validatorSchemaReader.ReadToEnd());
        }

        /// <summary>
        /// Performs a Json schema based validation.
        /// </summary>
        /// <param name="filePath">The path to the Json file to validate.</param>
        /// <returns>Whether the validation was successful.</returns>
        public static bool ValidateJson(string filePath)
        {
            var validationIssues = new List<ValidationIssue>();

            if (!File.Exists(filePath))
            {
                validationIssues.Add(new ValidationIssue(ValidationIssueType.Error, "The provided input file does not exist"));
            }
            else
            {
                JObject jObject = JObject.Parse(File.ReadAllText(filePath));

                if (!jObject.IsValid(schema, out IList<ValidationError> validationErrors))
                {
                    validationIssues.AddRange(
                        validationErrors.Select(e => new ValidationIssue(ValidationIssueType.Error, e.Message))
                                        .ToList());
                }
            }

            return ValidationHelper.RegisterValidationIssues(validationIssues);
        }

        /// <summary>
        /// Gets input data based on Json.
        /// </summary>
        /// <param name="filePath">The path to the Json file to get the input from.</param>
        /// <returns>The result of the operation.</returns>
        public static DataResult<ICalculationInput> GetInputDataFromJson(string filePath)
        {
            if (!File.Exists(filePath))
            {
                EventRegistry.Register(new Event("The provided input file does not exist", EventType.Error));
            }
            else
            {
                var jsonSerializer = new JsonSerializer();
                jsonSerializer.Converters.Add(new JsonInputCalculationDataConverter());
                jsonSerializer.Converters.Add(new JsonInputLocationDataConverter());

                try
                {
                    JObject jObject = JObject.Parse(File.ReadAllText(filePath));

                    return JsonInputAdapter.AdaptJsonInputData(jObject.ToObject<JsonInputData>(jsonSerializer));
                }
                catch (Exception e)
                {
                    EventRegistry.Register(new Event("An unhandled error occurred while composing calculation data" +
                                                     $" from the Json input. See stack trace for more information:\n{e.Message}",
                                                     EventType.Error));
                }
            }

            return new DataResult<ICalculationInput>(EventRegistry.Flush());
        }
    }
}